# フィーチャ レイヤーの使用

フィーチャ レイヤーは、GIS でフィーチャを操作するための主要な概念です。
インポート、エクスポート、分析、編集などにフィーチャ レイヤーを使用することができます。
また、フィーチャ レイヤーをマップに追加して可視化したりすることもできます。

フィーチャ レイヤーは、GIS にデータを公開することで作成されます。フィーチャ レイヤーは、フィーチャ レイヤー コレクションの layers プロパティで取得することができます。

## フィーチャ レイヤーへのアクセス

フィーチャ レイヤー を検索したり、アイテム ID を使用することでフィーチャ レイヤーにアクセスすることができます。

## フィーチャ レイヤーの検索

search() メソッドの item_type パラメーターを 'Feature Layer' または 'Feature Layer Collection'と指定することで、GIS 上でフィーチャ レイヤーを検索することができます。

以下の例では、実際にフィーチャ レイヤーを検索しています。

In [127]:
from arcgis.gis import GIS
gis = GIS() 

In [131]:
# タイトルが 'USA major cities' のフィーチャ レイヤーの検索 
search_results = gis.content.search('title: USA Major Cities',
                                    'Feature Layer')
search_results
# 検索結果の1番目のアイテムにアクセス
major_cities_item = search_results[0]

major_cities_item

<Item title:"USA Major Cities" type:Feature Layer Collection owner:esri_dm>

上記の major_cities_item は フィーチャ レイヤー コレクションです。layers プロパティにアクセスすると FeatureLayer オブジェクトのリストが表示されます。

In [71]:
major_cities_layers = major_cities_item.layers
major_cities_layers

[<FeatureLayer url:"https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/USA_Major_Cities/FeatureServer/0">]

## アイテム ID を使ったフィーチャ レイヤーへのアクセス

get() メソッドでアイテム ID を使用してフィーチャ レイヤーにアクセスすることもできます。

In [72]:
freeways = gis.content.get('91c6a5f6410b4991ab0db1d7c26daacb')
freeways

<Item title:"USA Freeway System" type:Feature Layer Collection owner:esri_dm>

同様に layers プロパティにアクセスすると FeatureLayer オブジェクトのリストが表示されます。このアイテムは2つのレイヤーを持っています。

In [73]:
freeways.layers 

[<FeatureLayer url:"https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/USA_Freeway_System/FeatureServer/1">,
 <FeatureLayer url:"https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/USA_Freeway_System/FeatureServer/2">]

以下では、レイヤーをループしその名前を表示しています。

In [74]:
for lyr in freeways.layers:
    print(lyr.properties.name)

USA Freeway System (over 1:500k)
USA Freeway System (below 1:500k)


## フィーチャ サービスからフィーチャ レイヤーとテーブルにアクセス

フィーチャ サービスは、フィーチャ レイヤーとテーブルを提供します。

以下に示すように、arcgis.features.FeatureLayerCollection でフィーチャ サービスの URL を使用してフィーチャ レイヤー コレクション作成することができます。

In [75]:
from arcgis.features import FeatureLayerCollection

In [76]:
fs_url = 'http://sampleserver3.arcgisonline.com/ArcGIS/rest/services/SanFrancisco/311Incidents/FeatureServer'
sanfran = FeatureLayerCollection(fs_url)
sanfran

<FeatureLayerCollection url:"http://sampleserver3.arcgisonline.com/ArcGIS/rest/services/SanFrancisco/311Incidents/FeatureServer">

フィーチャ レイヤー コレクション内のレイヤーとテーブルは、それぞれlayerプロパティとtableプロパティを使用してアクセスすることができます。

In [77]:
sanfran.layers

[<FeatureLayer url:"http://sampleserver3.arcgisonline.com/ArcGIS/rest/services/SanFrancisco/311Incidents/FeatureServer/0">]

In [11]:
sanfran.tables

[<Table url:"http://sampleserver3.arcgisonline.com/ArcGIS/rest/services/SanFrancisco/311Incidents/FeatureServer/1">]

## フィーチャ レイヤーの URL からフィーチャ レイヤーにアクセス

フィーチャ レイヤーは、REST エンドポイントの URL を使ってアクセスすることもできます。

In [78]:
from arcgis.features import FeatureLayer

In [79]:
lyr_url = 'http://sampleserver3.arcgisonline.com/ArcGIS/rest/services/SanFrancisco/311Incidents/FeatureServer/0'

layer = FeatureLayer(lyr_url)
layer

<FeatureLayer url:"http://sampleserver3.arcgisonline.com/ArcGIS/rest/services/SanFrancisco/311Incidents/FeatureServer/0">

<a id="featurelayer-properties"></a>
FeatureLayerオブジェクトのプロパティについて詳しく見ていきましょう。先ほど作成した major_cities_layers を使用します。

In [82]:
feature_layer = major_cities_layers[0]
feature_layer

<FeatureLayer url:"https://services.arcgis.com/P3ePLMYs2RVChkJx/arcgis/rest/services/USA_Major_Cities/FeatureServer/0">

FeatureLayer の properties フィールドは、プロパティを辞書型で表現します。

In [81]:
feature_layer.properties.extent

{
  "xmin": -17608123.3895845,
  "ymin": 2237818.89912024,
  "xmax": -7656889.54512499,
  "ymax": 9568526.08684907,
  "spatialReference": {
    "wkid": 102100,
    "latestWkid": 3857
  }
}

以下のように個々のプロパティにアクセスすることもできます。

In [83]:
feature_layer.properties.capabilities

'Query'

例えばレンダリングの情報は drawingInfo プロパティからアクセスできます。

In [84]:
feature_layer.properties.drawingInfo.renderer.type

'classBreaks'

<a id="querying-layers"></a>
## フィーチャ レイヤーのクエリ
クエリは、FeatureLayerオブジェクトに対して実行できる操作です。クエリを書くためには、レイヤーに存在するフィールドの名前を知る必要があります。以下では fields プロパティを呼び出しています。

In [85]:
for f in feature_layer.properties.fields:
    print(f['name'])

FID
NAME
CLASS
ST
STFIPS
PLACEFIPS
CAPITAL
POP_CLASS
POPULATION
POP2010
WHITE
BLACK
AMERI_ES
ASIAN
HAWN_PI
HISPANIC
OTHER
MULT_RACE
MALES
FEMALES
AGE_UNDER5
AGE_5_9
AGE_10_14
AGE_15_19
AGE_20_24
AGE_25_34
AGE_35_44
AGE_45_54
AGE_55_64
AGE_65_74
AGE_75_84
AGE_85_UP
MED_AGE
MED_AGE_M
MED_AGE_F
HOUSEHOLDS
AVE_HH_SZ
HSEHLD_1_M
HSEHLD_1_F
MARHH_CHD
MARHH_NO_C
MHH_CHILD
FHH_CHILD
FAMILIES
AVE_FAM_SZ
HSE_UNITS
VACANT
OWNER_OCC
RENTER_OCC


query() メソッドには多くのパラメーターがあり、結果の絞り込みを行うことができます。
例えば、2010年の人口が100万人を超える都市をすべて選択し、人口に関連するフィールドのみを取得してみましょう。

In [86]:
query_result1 = feature_layer.query(where='POP2010>1000000', 
                                    out_fields='WHITE,BLACK,MULT_RACE,HISPANIC')
len(query_result1.features)

9

In [87]:
query_result1.fields

[{'name': 'FID',
  'type': 'esriFieldTypeOID',
  'alias': 'FID',
  'sqlType': 'sqlTypeInteger',
  'domain': None,
  'defaultValue': None},
 {'name': 'WHITE',
  'type': 'esriFieldTypeInteger',
  'alias': 'WHITE',
  'sqlType': 'sqlTypeInteger',
  'domain': None,
  'defaultValue': None},
 {'name': 'BLACK',
  'type': 'esriFieldTypeInteger',
  'alias': 'BLACK',
  'sqlType': 'sqlTypeInteger',
  'domain': None,
  'defaultValue': None},
 {'name': 'MULT_RACE',
  'type': 'esriFieldTypeInteger',
  'alias': 'MULT_RACE',
  'sqlType': 'sqlTypeInteger',
  'domain': None,
  'defaultValue': None},
 {'name': 'HISPANIC',
  'type': 'esriFieldTypeInteger',
  'alias': 'HISPANIC',
  'sqlType': 'sqlTypeInteger',
  'domain': None,
  'defaultValue': None}]

return_count_only パラメーターを True に設定することで該当するフィーチャの数のみを表示することができます。

In [88]:
feature_layer.query(where='POP2010>1000000', return_count_only=True)

9

## 異なる空間参照を使用したフィーチャのクエリ

In [89]:
query_result1.spatial_reference

{'wkid': 102100, 'latestWkid': 3857}

デフォルトでは、クエリの結果はもとのレイヤーと同じ空間参照になります。ただし、`out_sr` パラメータを使用して、結果を任意の空間参照に再投影することができます。

上記の例では、'Web Mercator' の wkid:3857 のデータを取得しました。座標がどのように表示されるかを見てみましょう。

In [90]:
query_result1.features[0].geometry

{'x': -9756835.705284344, 'y': 5124572.044517264}

座標は投影座標系で表示されています。このデータを緯度経度で表示したい場合は、 out_sr を wkid:4326 に変更することで表示することができます。

<a id = "featureset-properties"></a>
## フィーチャ セット プロパティ
FeatureSet オブジェクトは query() によって返されるオブジェクトです。FeatureSet オブジェクトには、アクセス対象のフィーチャに関する有用な情報を識別するのに役立つ多くの便利なプロパティが含まれています。

その中で重要なプロパティの一つが先ほども使用した spatial_reference です。以下では、先ほどのクエリと同じ query_result1 を用いています。

In [91]:
query_result1.spatial_reference

{'wkid': 102100, 'latestWkid': 3857}

FeatureSet オブジェクトは，pandas のデータフレームオブジェクトとしてフィーチャにアクセスすることができです．sdf プロパティは、データフレームオブジェクトを返します。

In [92]:
query2 = feature_layer.query(where="POP2010 > 1000000")
query2.sdf

FID          NAME CLASS  ST STFIPS PLACEFIPS CAPITAL  POP_CLASS  \
0    63       Chicago  city  IL     17   1714000                 10   
1   415   Los Angeles  city  CA     06   0644000                 10   
2   550     San Diego  city  CA     06   0666000                 10   
3   910       Phoenix  city  AZ     04   0455000   State         10   
4  2174      New York  city  NY     36   3651000                 10   
5  2987  Philadelphia  city  PA     42   4260000                 10   
6  3592        Dallas  city  TX     48   4819000                 10   
7  3641       Houston  city  TX     48   4835000                 10   
8  3726   San Antonio  city  TX     48   4865000                 10   

   POPULATION  POP2010  ...  MARHH_NO_C  MHH_CHILD  FHH_CHILD  FAMILIES  \
0     2781116  2695598  ...      168435      28332     116555    576793   
1     3986442  3792621  ...      245056      46908     118939    807326   
2     1397856  1307402  ...      105134      12062      33865    285221   
3     1601381  1445632  ...      102047      23353      54452    330762   
4     8691599  8175133  ...      566484      77061     344755   1850221   
5     1587761  1526006  ...       95310      18486      86946    340354   
6     1323651  1197816  ...       76902      14512      50242    265538   
7     2333285  2099451  ...      145478      26039      85014    481570   
8     1442472  1327407  ...      100314      16067      56901    318043   

   AVE_FAM_SZ  HSE_UNITS  VACANT  OWNER_OCC  RENTER_OCC  \
0        3.40    1194337  148777     469562      575998   
1        3.53    1413995   95827     503863      814305   
2        3.28     516033   32941     233158      249934   
3        3.42     590149   75343     296742      218064   
4        3.32    3371062  261278     962892     2146892   
5        3.20     670171   70435     324536      275200   
6        3.42     516639   58582     201880      256177   
7        3.38     892646  110003     355236      427407   
8        3.34     524246   44604     271070      208572   

                                               SHAPE  
0  {"x": -9756835.705284344, "y": 5124572.0445172...  
1  {"x": -13165820.99345245, "y": 4035892.5853389...  
2  {"x": -13040584.546974456, "y": 3858240.841215...  
3  {"x": -12476005.927345268, "y": 3954668.990210...  
4  {"x": -8238770.183515309, "y": 4969744.1655956...  
5  {"x": -8366882.6254769275, "y": 4858876.345453...  
6  {"x": -10775254.59703981, "y": 3865959.6760583...  
7  {"x": -10616262.315905476, "y": 3472578.411494...  
8  {"x": -10964134.924024679, "y": 3429663.909911...  

[9 rows x 50 columns]

データフレームとしてフィーチャにアクセスすることで、統計的にデータを分析しやすくなります。

## クエリの結果からフィーチャにアクセス

In [93]:
query_geographic = feature_layer.query(where='POP2010 > 1000000', out_sr='4326')
query_geographic.features[0].geometry

{'x': -87.64714638699998, 'y': 41.75648806100002}

最初の FeatureLayer オブジェクトに対して query() メソッドを実行して FeatureSet を取得します。このレイヤーの最初の10個のフィーチャをクエリしてアクセスしてみましょう。

In [94]:
major_cities_l1 = major_cities_layers[0]
major_cities_l1_fset = major_cities_l1.query(where= 'FID < 11')
type(major_cities_l1_fset)

arcgis.features.feature.FeatureSet

上記の FeatureSet オブジェクトの features プロパティにアクセスすると、個々のポイントのフィーチャが返されます。

In [95]:
major_cities_l1_features = major_cities_l1_fset.features
len(major_cities_l1_features)

10

<a id="feature-properties"></a>
## フィーチャのジオメトリと属性へのアクセス
フィーチャは空間情報を細かく表現したものであり、ジオメトリや属性を持っています。

1つ目のフィーチャのジオメトリと属性を表示してみましょう。

In [96]:
major_cities_l1_features[0].geometry

{'x': -12462673.723706165, 'y': 5384674.994080178}

In [97]:
major_cities_l1_features[0].attributes

{'FID': 1,
 'NAME': 'Ammon',
 'CLASS': 'city',
 'ST': 'ID',
 'STFIPS': '16',
 'PLACEFIPS': '1601990',
 'CAPITAL': ' ',
 'POP_CLASS': 6,
 'POPULATION': 15181,
 'POP2010': 13816,
 'WHITE': 13002,
 'BLACK': 73,
 'AMERI_ES': 67,
 'ASIAN': 113,
 'HAWN_PI': 9,
 'HISPANIC': 884,
 'OTHER': 307,
 'MULT_RACE': 245,
 'MALES': 6750,
 'FEMALES': 7066,
 'AGE_UNDER5': 1468,
 'AGE_5_9': 1503,
 'AGE_10_14': 1313,
 'AGE_15_19': 1058,
 'AGE_20_24': 734,
 'AGE_25_34': 2031,
 'AGE_35_44': 1767,
 'AGE_45_54': 1446,
 'AGE_55_64': 1136,
 'AGE_65_74': 665,
 'AGE_75_84': 486,
 'AGE_85_UP': 209,
 'MED_AGE': 29.6,
 'MED_AGE_M': 28,
 'MED_AGE_F': 30.8,
 'HOUSEHOLDS': 4476,
 'AVE_HH_SZ': 3.05,
 'HSEHLD_1_M': 457,
 'HSEHLD_1_F': 648,
 'MARHH_CHD': 1618,
 'MARHH_NO_C': 1131,
 'MHH_CHILD': 106,
 'FHH_CHILD': 335,
 'FAMILIES': 3352,
 'AVE_FAM_SZ': 3.61,
 'HSE_UNITS': 4747,
 'VACANT': 271,
 'OWNER_OCC': 3205,
 'RENTER_OCC': 1271}

<a id="features-from-a-feature-collection"></a>
# フィーチャ コレクションの使用

フィーチャ レイヤーと同様に、[フィーチャ コレクション](http://doc.arcgis.com/en/arcgis-online/reference/feature-collections.htm)もフィーチャを保存するために使用することができます。フィーチャ コレクションでは、フィーチャ データを提供するためのサービスは作成されず、アイテムと一緒に json データとして保存されます。フィーチャ コレクションは、レイヤーとしてマップに追加したり、フィーチャ分析ツールへの入力として渡したり、フィーチャ データのクエリを行ったりすることができます。

フィーチャ コレクションはアイテムとして共有されます。フィーチャ コレクションのアイテムは、item_type に「Feature Collection」を指定して検索することができます。

例として、Learn_ArcGIS が公開しているフィーチャコレクションを検索してみましょう。

In [114]:
#search_fc = gis.content.search("title:AVL_Direct_FC", item_type='Feature Collection')
search_fc = gis.content.search("owner:Learn_ArcGIS", item_type='Feature Collection')
fc_item = search_fc[0]
fc_item

<Item title:"Emergency Shelters" type:Feature Collection owner:Learn_ArcGIS>

フィーチャ コレクションの layers プロパティにアクセスすると、FeatureCollection オブジェクトのリストを返します。

In [115]:
fc_item.layers

[<FeatureCollection>]

In [116]:
fc = fc_item.layers[0]

FeatureCollection オブジェクトで query() メソッドを呼び出して FeatureSet オブジェクトを取得することができます。

In [119]:
fset = fc.query()

FeatureSet オブジェクトを取得したら、features プロパティにアクセスして Feature オブジェクトのリストを取得することができます。

In [120]:
features = fset.features
features[0].geometry

{'x': -17265172.6773,
 'y': 2240027.716499999,
 'spatialReference': {'wkid': 102100, 'latestWkid': 3857}}